### PR TITLE
[Spark] Identity Columns Value Generation (without MERGE support)

### PR DESCRIPTION
**Pull request by c27kwan**
#### Which Delta project/connector is this regarding?
- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
This PR is part of https://github.com/delta-io/delta/issues/1959

In this PR, we enable basic ingestion for Identity Columns. 
* We use a custom UDF `GenerateIdentityValues` to generate values when not supplemented by the user. 
* We introduce classes to help update and track the high watermark of identity columns. 
* We also do some cleanup/ improve readability for ColumnWithDefaultExprUtils

Note: This does NOT enable Ingestion with MERGE INTO yet. That will come in a follow up PR, to make this easier to review.

## How was this patch tested?
We introduce a new test suite IdentityColumnIngestionSuite.

## Does this PR introduce _any_ user-facing changes?
No.

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnIngestionSuite.scala [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+
+    // Ingest using "INSERT OVERWRITE ... SELECT ...".
+    val insertOverwriteSelect = Value
+
+
+    // Ingest using streaming query.
+    val streaming = Value
+
+    // Ingest using MERGE INTO ... WHEN NOT MATCHED INSERT
+    val mergeInsert = Value
```
Comment (c27kwan) [inline spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnIngestionSuite.scala:75]: This is part about merge is not used yet, it's just easier to commit the test infrastructure in this PR.

File: spark/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala [hunk 2]
```diff
@@ -260,6 +261,24 @@ case class WriteIntoDelta(
               sparkSession.conf.get(DeltaSQLConf.REPLACEWHERE_DATACOLUMNS_WITH_CDF_ENABLED) &&
               cdcExistsInRemoveOp) {
             var dataWithDefaultExprs = data
+            // Add identity columns if they are not in `data`.
+            // Column names for which we will track identity column high water marks.
+            val trackHighWaterMarks = scala.collection.mutable.Set[String]()
```
Comment (larsk-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala:266]:
```suggestion
            val trackHighWaterMarks = scala.collection.mutable.Set.empty[String]
```
I'd import `scala.collection.mutable`.

File: spark/src/main/scala/org/apache/spark/sql/delta/files/TransactionalWrite.scala [hunk 3]
```diff
@@ -450,13 +465,21 @@ trait TransactionalWrite extends DeltaLogging { self: OptimisticTransactionImpl
           partitionColumns = partitioningColumns,
           bucketSpec = None,
           statsTrackers = optionalStatsTracker.toSeq
-            ++ statsTrackers,
+            ++ statsTrackers
+            ++ identityTrackerOpt.toSeq
+          ,
```
Comment (larsk-db) [inline spark/src/main/scala/org/apache/spark/sql/delta/files/TransactionalWrite.scala:470]:
```suggestion
            ++ identityTrackerOpt.toSeq,
```

**Review by larsk-db — APPROVED** [review]
Just some nits.

**Pull request merged**